### PR TITLE
More miro merging

### PR DESCRIPTION
**Pull request by jamieparkinson**
This achieves the following:

_In the transformer:_
- Removes the restriction on material type for Sierra's Miro mergeCandidates
- Remove the requirement that these mergeCandidates are 1-to-1 (ie, merge any number of linked Miro works)

_In the merger:_
- Suppresses any "historical library" (`L`/`M`) Miro works **that are not linked with a Sierra work** and does not create images from them.
- Otherwise, maintains existing behaviour by only merging in items/identifiers and redirecting Sierra/Miro works if there is **1** Miro work and the Sierra work has a `workType` of `Picture`, `Digital Image`, or `3D Object`.

Resolves https://github.com/wellcomecollection/platform/issues/4565

### CODE REVIEW

File: common/internal_model/src/test/scala/uk/ac/wellcome/models/work/generators/IdentifiersGenerators.scala [hunk 1]
```diff
@@ -52,8 +54,22 @@ trait IdentifiersGenerators extends RandomStrings {
       identifierType = IdentifierType("isbn")
     )
 
+  private val MIRO_ID_PREFIXES: Seq[Char] = Seq(
```
Comment (jamesgorrie) [inline common/internal_model/src/test/scala/uk/ac/wellcome/models/work/generators/IdentifiersGenerators.scala:57]: Why all the big letters?

Comment (jamieparkinson) [reply common/internal_model/src/test/scala/uk/ac/wellcome/models/work/generators/IdentifiersGenerators.scala:57]: Because it's a constant and I've written too much C and it has rotted my brain - I'll change it to `MiroIdPrefixes` as per https://docs.scala-lang.org/style/naming-conventions.html#constants-values-variable-and-methods

Comment (warrd) [reply common/internal_model/src/test/scala/uk/ac/wellcome/models/work/generators/IdentifiersGenerators.scala:57]: Fwiw I haven't seen `ConstantsLikeThis` once in our code base. I tend to just use `standardCamelCaseForConstants` (there are loads of examples, they are all over the place), and don't find it particularly confusing / misleading, but can understand if people not into it.

Comment (jamesgorrie) [reply common/internal_model/src/test/scala/uk/ac/wellcome/models/work/generators/IdentifiersGenerators.scala:57]: I was just writing what @warrd said, if we are going to come up with style rules, we need to be able to lint them otherwise these conversations become often and boring

Comment (jamieparkinson) [reply common/internal_model/src/test/scala/uk/ac/wellcome/models/work/generators/IdentifiersGenerators.scala:57]: OK sure - I certainly can't be bothered with linting since we have scalafmt. That said this conversation is evidence that we do have style rules, just undocumented ones that are different to the ones that someone who was looking for Scala style guidance would find, no? 😛

Comment (warrd) [reply common/internal_model/src/test/scala/uk/ac/wellcome/models/work/generators/IdentifiersGenerators.scala:57]: I would probably agree with that statement, yep :see_no_evil:

Comment (warrd) [reply common/internal_model/src/test/scala/uk/ac/wellcome/models/work/generators/IdentifiersGenerators.scala:57]: But I kind of think that is unavoidable for any team of devs, as no guidelines covers everything. It might be easier to follow the official scala style guide for these sort of cases, but starting at this point would probably lead to lots of inconsistencies (not that in this case I think it particularly matters if we were slightly inconsistent tho)

File: common/internal_model/src/main/scala/uk/ac/wellcome/models/work/internal/InvisibilityReason.scala [hunk 2]
```diff
@@ -8,4 +8,5 @@ object InvisibilityReason {
   case class CopyrightNotCleared(info: String) extends InvisibilityReason
   case class SourceFieldMissing(info: String) extends InvisibilityReason
   case class InvalidValueInSourceField(info: String) extends InvisibilityReason
+  case class UnlinkedHistoricalLibraryMiro() extends InvisibilityReason
```
Comment (warrd) [inline common/internal_model/src/main/scala/uk/ac/wellcome/models/work/internal/InvisibilityReason.scala:11]: Any reason this needs to be a `case class` rather than a `case object`?

File: pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/rules/ItemsRule.scala [hunk 3]
```diff
@@ -15,33 +11,34 @@ import cats.data.NonEmptyList
...
-    } ++ findFirstLinkedDigitisedSierraWorkFor(target, sources)
+    val mergedSources = (
+      List(
+        mergeIntoCalmTarget,
+        mergeMetsIntoSingleItemSierraTarget,
+        mergeSingleMiroIntoSingleItemSierraTarget,
+        mergeIntoMultiItemSierraTarget)
+        .flatMap(_.mergedSources(target, sources)) ++
+        findFirstLinkedDigitisedSierraWorkFor(target, sources)
+    ).distinct
```
Comment (warrd) [inline pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/rules/ItemsRule.scala:41]: nit: the way this is indented / laid out makes it pretty hard to parse for me (dunno if this is scalafmt being a nuisance or not)

File: common/internal_model/src/main/scala/uk/ac/wellcome/models/work/internal/InvisibilityReason.scala (hunk 2 above)
Comment (jamieparkinson) [reply common/internal_model/src/main/scala/uk/ac/wellcome/models/work/internal/InvisibilityReason.scala:11]: Just for consistency with the others tbh

Comment (warrd) [reply common/internal_model/src/main/scala/uk/ac/wellcome/models/work/internal/InvisibilityReason.scala:11]: I don't think inconsistent, I think others are that way because have an `info` argument. Maybe we want an `info` string for this too?

Comment (jamieparkinson) [reply common/internal_model/src/main/scala/uk/ac/wellcome/models/work/internal/InvisibilityReason.scala:11]: Yeah, I couldn't find anything that would be useful `info` for this - the others use it to say (eg) what their source is. I'll make it an `object`

File: pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/rules/FieldMergeRule.scala [hunk 4]
```diff
@@ -45,20 +46,31 @@ trait FieldMergeRule {
   protected trait PartialRule {
     val isDefinedForTarget: WorkPredicate
     val isDefinedForSource: WorkPredicate
+    val isDefinedForSourceList: Seq[TransformedBaseWork] => Boolean =
```
Comment (jamesgorrie) [inline pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/rules/FieldMergeRule.scala:49]: For a rule to be applied:
* The `target` has to return `true` with the `isDefinedForTarget` predicate
* The `sources` has to return `true` with the `isDefinedForSourceList` predicate
* Only is some `sources` that return `true` with the `isDefinedForSource` predicate

And only the `isDefinedForSource` `sources` get sent to the rule.

Is that about right?

File: pipeline/transformer/transformer_sierra/src/main/scala/uk/ac/wellcome/platform/transformer/sierra/transformers/SierraMergeCandidates.scala [hunk 5]
```diff
@@ -70,37 +69,15 @@ object SierraMergeCandidates
...
-                "Single page Miro/Sierra work (secondary source)")
-            )
-          case _ => Nil
-        }
-      case _ => Nil
-    }
+    (matching089Ids(bibData) ++ matching962Ids(bibData)).distinct
+      .map {
+        miroMergeCandidate(_, "Miro/Sierra work")
+      }
```
Comment (jamesgorrie) [inline pipeline/transformer/transformer_sierra/src/main/scala/uk/ac/wellcome/platform/transformer/sierra/transformers/SierraMergeCandidates.scala:80]: I much prefer this process, it's so much easier to tell what's going on where.

File: pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/rules/FieldMergeRule.scala [hunk 6]
```diff
@@ -45,20 +46,31 @@ trait FieldMergeRule {
...
+        case head +: tail =>
           Some(rule(target, NonEmptyList(head, tail.toList)))
         case _ => None
       }
 
     def apply(target: UnidentifiedWork,
               source: TransformedBaseWork): Option[FieldData] =
       apply(target, List(source))
+
+    def mergedSources(
```
Comment (jamesgorrie) [inline pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/rules/FieldMergeRule.scala:67]: Maybe a little comment as to what this method is for?

File: pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/rules/FieldMergeRule.scala (hunk 4 above)
Comment (jamieparkinson) [reply pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/rules/FieldMergeRule.scala:49]: That is exactly right
